components.go

```diff
@@ -263,11 +264,28 @@ func (cb *ComponentBase) SetDuration(d time.Duration) error {
 	return errors.New("start or end not yet defined")
 }
 
-func (cb *ComponentBase) GetEndAt() (time.Time, error) {
-	return cb.getTimeProp(ComponentPropertyDtEnd, false)
-}
-
-func (cb *ComponentBase) getTimeProp(componentProperty ComponentProperty, expectAllDay bool) (time.Time, error) {
+// GetEndAt gets the date and time that the component finishes at, must not be an all day event.
+// Warning: by default uses "Local" timezone -- To overwrite it, supply the desired timezone as one of the optional arguments. Ie, GetEndAt(time.UTC)
+// If you have a property that specifies a timezone then that is used instead.
+func (cb *ComponentBase) GetEndAt(ops ...any) (time.Time, error) {
```

Comment:
This is a breaking change. Anyone who uses types embedding ComponentBase through an interface will have to update the method signature in their interface. If any module has two or more dependencies which do this and they don't all have the same interface, then it will be impossible to build that module.

Reply:
I'd be quite happy adding a new constructor function like if we want to use the functional options pattern to resolve this.
```go
type Option func(*Calendar)

func ParseCalendarOpts(r io.Reader, opts ...Option) (*Calendar, error)
```

Reply:
The constructor is unlikely to be in any interface as it doesn't have a receiver. 

Should I provide a proxy struct for quick adaptation. (This library is still pre v1 which is not something I want to say but there are a couple of changes I might need to make that are breaking. If only there was a way of finding out who was using what...)
```
type NoOpComponentBase struct {
   *ComponentBase
}

func (nocb *NoOpCompontentBase) GetEndAt() (time.Time, error) {
  return nocb.ComponentBase.GetEndAt()
}
```

Which could be used as a dropin replacement? + Documentation. A little more work when they upgrade but hopefully something I could put in the function documentation.

Does that work? Or will I need to:
```
func (cb *ComponentBase) GetEndAt() (time.Time, error) {
}

func (cb *ComponentBase) GetEndAtWithOptions(ops ...any) (time.Time, error) {
}
```
?

Are we talking specifically about `GetEndAt` or the constructor?

Reply:
You are right, you can make breaking changes without breaking the modules promise as you're v0.x. But you are easily the most popular (used by 327 public repos) and most mature Go ICS library. If it were my inbox I'd avoid breaking changes where practical. Otherwise jumping right to v2 to make breaking changes would be better.

I'm not sure how you would wire `NoOpComponentBase` in? As `ComponentBase` is exported, changing the type that things like `VAlarm` embed is also a breaking change.

I still think choosing a fallback location when the `Calendar` is instantiated is pretty clean, and a new constructor a pretty clean compatible change to do that. That's my opinion.

The reason for making option a type is so its clear what you provide, in my example it would use:
```go
// InLocation provides an Option to set the fallback location. time.Time values
// for events in the calendar which have no explicit location, such as all day
// events, will use this location. The default fallback location is time.Local.
func InLocation(l time.Location) Option {
    return func(c *Calendar) {
        c.fallbackLocation = l
    }
}
```

So that you can write:
```go
cal, err := ics.ParseCalendar(r, ics.InLocation(userLocation))
```

and according to Rob Pike *[interface{} says nothing](https://go-proverbs.github.io/)*.

Reply:
I've replied in PR:
https://github.com/arran4/golang-ical/pull/107#issuecomment-2425459018